# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:**

**Fecha límite de entrega 📆:** Lunes 29 de Marzo.

**Tiempo estimado de dedicación:**




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 5 lineas** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones.**



> El análisis del lenguaje humano es complejo porque el lenguaje humano es **ambiguo**: por ejemplo, las oraciones *comí empanadas con pebre* y *comí empanadas con Luis* son formalmente muy similares, ambas solo se diferencian en las palabras *pebre* y *Luis*, sin embargo, tienen significados completamente distintos; es decir, dichas oraciones son sintácticamente muy similares pero semánticamente distintas. Lo anterior conlleva a que analizar el lenguaje humano es más complejo que solamente estudiar *formalmente* las palabras y oraciones . Asimismo, relacionado con lo anterior, el análisis del lenguaje humano también es complejo porque el lenguaje **depende del contexto**: la misma palabra puede tener un significado completamente distinto dependiendo del lugar, tiempo y/o grupo humano.

**Pregunta 2 (0.3 puntos): ¿Cuál es la diferencia entre Natural Language Processing (NLP) y Computational Linguistics?**

> Natural Language Processing es una disciplina que se encarga de resolver tareas particulares que involucran el lenguaje humano y, en ese sentido, en NLP se buscan métodos y algoritmos que *solucionen* problemas que involucren lenguaje humano. Por el otro lado, Computational Linguistics es una disciplina que estudia la *linguistica* mediante la ayuda de herramientas computacionales y, en ese sentido, en la disciplina Computational Linguistics el objeto de estudio principal es el lenguaje en sí mismo.


**Pregunta 3 (0.3 puntos): ¿Qué es el Machine Learning supersivado? Donde podría aplicar este tipo de Aprendizaje**

> Machine Learning supervisado es sub-área de la disciplina de Machine Learning que se define por utilizar datasets *etiquetados* para entrenar algoritmos para clasificar data o predecir *outputs* de manera precisa. Por ejemplo, en NLP se pueden clasificar textos utilizando textos pre-etiquetados: supóngase que se tiene un conjunto de textos etiquetados en relación al tipo de texto (textos políticos, textos deportivos y textos científicos); dado dichos textos, se puede utilizar algún algoritmo de Machine Learning supervisado para que, dado un texto nuevo no etiquteado, se etiquete dicho texto en alguna de las categorías de los textos iniciales.


**Pregunta 4 (0.3 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.**

> Para utlizar Machine Learning clásico en NLP se necesita establecer una representación particular para la data, es decir, dado un texto, se necesita predefinir la *mejor* forma de representar dicho texto. Por el otro lado, en Deep learning no se necesita establecer una representación particular para la data, el algoritmo encuentra por sí solo la mejor representación para los datos. Por ejemplo, en el caso de clasificar tweets según sentimiento, en un enfoque clásico por Machine Learning se tiene que establecer de qué forma se va a representar el tweet. Se puede crear un vector que contenga información relevante sobre un el tweet: una columna puede ser el número de negaciones individuales del tweet, otra columna puede ser el número de palabras en mayúsculas, etcétera. Notése que existen muchas representaciones y definir cuál es la mejor es una tarea compleja. En cambio, en el caso de Deep learning, el algoritmo mismo será el que escoga qué elementos del tweet son los más relevante para definir su *sentimiento*.

 **Pregunta 5 (0.3 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

> tf-idf scoring model

**Pregunta 6 (0.3 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.**

> La idea del jefe podría ser correcta ya que, en general, en un texto las palabras más comunes (palabras como de, y, el, la, que, un, por, para) no dicen tanto sobre la semántica de dicho texto (y, en este caso, se quiere capturar la semántica del texto porque se quiere encontrar los *sentimientos* de dicho texto). Por ejemplo, en las oraciones "*de anoche que estoy enojado*" y "*es inevitable que esté feliz si paso de estar nublado a soleado*" es claro que las palabras *enojado* y *feliz* son las palabrás que capturan los sentimientos de dichos textos (y las palabras *de* y *que* no entregan información alguna). Sin embargo, en otros casos, eliminar palabras podría ser una mala idea. Por ejemplo, en las oraciones "*la fila del supermercado es muy pequeña*" y "*el sueldo que me pagaron terminó siendo bastante pequeño*" el sentimiento del texto no está tanto en una o en un par de palabras sino en el contexto del texto: la palabra *pequeña* puede ser positiva o negativa. En dichos casos, eliminar palabras no necesariamente es una buena opción para obtener una mejor clasificación.

**Pregunta 7 (0.3 puntos) Nombre las ventajas y desventajas de utilizar un modelo Bag of Words y TFIDF**

> Algunos ventajas de utilizar un modelo Bag of Words y TFIDF son las siguientes:
> - Es un modelo simple tanto en términos de espacio (el vocabulario generado es a lo más del tamaño del vocabulario de un lenguaje humano y la representación vectorial de los textos permite almacenar dichos vectores de forma *sparse*) como en términos de complejidad (crear el diccionario es proporcionalmente lineal al tamaño de todos los textos y las operaciones sobre las representaciones vectoriales de los textos son proporcionales al tamaño del vocabulario).
> - Permite representar de forma simple los textos y, por tanto, las *operaciones* entre dos textos es simple. Por ejemplo, dada dos representaciones de textos, encontrar qué términos se hayan en ambos textos es una operación sumamente simple.
> - Relacionado con el punto anterior, el modelo permite buscar términos en textos de forma simple, es decir, responder a la pregunta si un término se encuentra o no en un texto en particular es una tarea fácil.
>
>
> Por el otro lado, algunas desventajas:
> - Cuando se representan un texto como vector y se almacena la información respecto a qué palabras se hayan en dicho texto, se pierde la información *espacial* de dicho texto, es decir,la representación vectorial no contiene información sobre *dónde* se encontraban las palabras presentes en el texto.
> Es *explotable* en el sentido de que se puede engañar al modelo  usando términos que se refieren a una palabra pero no de forma explícita. Por ejemplo, si se quisiera que la palabra *política* no sea identifica en el texto, se podría cambiar dicha palabra por las "P O L I T I C A", "p0l1t1c4", "politika", entre otras infinitas variaciones de dicha palabra. 


## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Typos*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [1]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [2]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [3]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

In [4]:
try:
    from google.colab import drive

    drive.mount("/content/drive")
    path = # Rellenar con la ruta del drive donde alojan la tarea
except: 
    print('Ignorando conexión drive-colab')

SyntaxError: invalid syntax (2601773408.py, line 5)

` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver que tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [17]:
text = codecs.open('marcianeke.txt', 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [18]:
from functools import reduce
def get_tokens(text):
    tokens = text.split('\n')
    tokens = list(map(lambda x: x.strip().split(' '), tokens))
    tokens = reduce(lambda x, y: x + y, tokens)
    tokens = reduce(lambda l, el: l + re.split(r"(\w+)", el), tokens, [])
    tokens = list(filter(lambda x: x != '', tokens))
    return tokens


In [19]:
tokens = get_tokens(text)
tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

 > Se supuso que un token es una letra, una palabra o un símbolo. En particular, se intentó seguir a misma forma de *tokenizar* que el ejemplo dado.

In [20]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

Nótese que se obtuvo el mismo resultado que el ejemplo. La siguiente línea de código evidencia lo mencionado.

In [21]:
print(nltk_tokens == tokens)

True


` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [24]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los typos de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [25]:
def get_vocab(dataset):
    typos = list(map(get_tokens, dataset))
    typos = reduce(lambda x, y: x + y, typos, [])
    typos = set(typos)
    return list(typos)

***Test:***

In [26]:
vocab = get_vocab(dataset)
vocab

['languages',
 'programming',
 'my',
 'favorite',
 'like',
 'Spanish',
 'language',
 'human',
 'is',
 'I']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

> Para las reglas asociadas a Stemming se utilizó el algoritmo de Porter con las siguientes reglas:
> - SSES $\rightarrow$ SS
> - IES $\rightarrow$ I
> - SS $\rightarrow$ SS
> - S $\rightarrow$
>
> Para reglas asociadas a Stopwords se utilizó la siguiente lista de stopwords: <a href="https://gist.github.com/sebleier/554280">NLTK's list of english stopwords</a>

In [31]:
def pre_processing(vocab):
    
    def porter(token):
        if token.endswith("sses"):
            token = re.sub("(sses)$", "ss", token)
        elif token.endswith("ies"):
            token = re.sub("(ies)$", "i", token)
        elif token.endswith("ss"):
            token = re.sub("(ss)$", "ss", token)
        elif token.endswith("s"):
            token = re.sub("(s)$", "", token)
        return token
    
    def is_stopword(token):
        stopwords = ["i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very", "s", "t", "can", "will", "just", "don", "should", "now"]
        return token in stopwords
    
    vocab = map(lambda tok: tok.lower(), vocab) # lower
    vocab = map(porter, vocab) # porter stemming
    vocab = filter(lambda x: not is_stopword(x), vocab) # filter stopwords
    vocab = set(vocab) # remove non-unique words
    return list(vocab)

In [32]:
vocab = pre_processing(vocab)
vocab

['programming', 'favorite', 'like', 'human', 'spanish', 'language']


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [33]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [34]:
def bag_of_words(dataset):
    vocab = get_vocab(dataset)
    
    def freq_list(document):
        freqs = {key: 0 for key in vocab}
        for token in get_tokens(document):
            freqs[token] += 1
        return [freqs[word] for word in vocab]
    
    frequencies = list(map(freq_list, dataset))
    dataset_names = map(lambda x: f"d{x}", range(len(dataset)))
    frame = pd.DataFrame(data=frequencies, index=dataset_names, columns=vocab)
    return frame

***Test:***

In [35]:
dataset_bow = bag_of_words(dataset)
dataset_bow

maulla  ladra  la  El  empieza  ladrar  gato  comida  maullar  se  a  y  \
d0       0      0   1   1        0       0     0       1        0   2  0  1   
d1       0      0   0   1        1       1     0       0        0   1  1  1   
d2       0      1   1   1        0       0     0       1        0   1  0  1   
d3       0      0   1   1        0       0     1       1        0   2  0  1   
d4       0      0   0   1        1       0     1       0        1   1  1  1   
d5       1      0   1   1        0       0     1       1        0   1  0  1   

    perro  despierta  duerme  después  come  
d0      1          0       1        1     1  
d1      1          1       0        1     0  
d2      1          0       0        1     1  
d3      0          0       1        1     1  
d4      0          1       0        1     0  
d5      0          0       0        1     1

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [36]:
def calc_tf(dataset_bow):
    maxs = dataset_bow.apply(max, 1)
    data = dataset_bow.apply(lambda vec: list(map(lambda x: x[0] / x[1],zip(vec, maxs))), 0)
    return data

***Test:***

In [37]:
tf = calc_tf(dataset_bow)
tf

maulla  ladra   la   El  empieza  ladrar  gato  comida  maullar   se    a  \
d0     0.0    0.0  0.5  0.5      0.0     0.0   0.0     0.5      0.0  1.0  0.0   
d1     0.0    0.0  0.0  1.0      1.0     1.0   0.0     0.0      0.0  1.0  1.0   
d2     0.0    1.0  1.0  1.0      0.0     0.0   0.0     1.0      0.0  1.0  0.0   
d3     0.0    0.0  0.5  0.5      0.0     0.0   0.5     0.5      0.0  1.0  0.0   
d4     0.0    0.0  0.0  1.0      1.0     0.0   1.0     0.0      1.0  1.0  1.0   
d5     1.0    0.0  1.0  1.0      0.0     0.0   1.0     1.0      0.0  1.0  0.0   

      y  perro  despierta  duerme  después  come  
d0  0.5    0.5        0.0     0.5      0.5   0.5  
d1  1.0    1.0        1.0     0.0      1.0   0.0  
d2  1.0    1.0        0.0     0.0      1.0   1.0  
d3  0.5    0.0        0.0     0.5      0.5   0.5  
d4  1.0    0.0        1.0     0.0      1.0   0.0  
d5  1.0    0.0        0.0     0.0      1.0   1.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [38]:
def calc_idf(dataset_bow):
    N = len(dataset_bow.index)
    n = dataset_bow.apply(lambda x: reduce(lambda x, y: x + (1 if y >= 1 else 0), x, 0), 0)
    idf = n.map(lambda n_i: np.log10(N/n_i))
    return dict(idf)

***Test:***

In [39]:
idf = calc_idf(dataset_bow)
idf

{'maulla': 0.7781512503836436,
 'ladra': 0.7781512503836436,
 'la': 0.17609125905568124,
 'El': 0.0,
 'empieza': 0.47712125471966244,
 'ladrar': 0.7781512503836436,
 'gato': 0.3010299956639812,
 'comida': 0.17609125905568124,
 'maullar': 0.7781512503836436,
 'se': 0.0,
 'a': 0.47712125471966244,
 'y': 0.0,
 'perro': 0.3010299956639812,
 'despierta': 0.47712125471966244,
 'duerme': 0.47712125471966244,
 'después': 0.0,
 'come': 0.17609125905568124}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [40]:
def calc_tf_idf(tf, idf):
    tfidf = tf.copy()
    for word in tf.columns:
        tfidf[word] = tfidf[word].map(lambda x: idf[word] * x)
    return tfidf

***Test.***

In [41]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

maulla     ladra        la   El   empieza    ladrar      gato    comida  \
d0  0.000000  0.000000  0.088046  0.0  0.000000  0.000000  0.000000  0.088046   
d1  0.000000  0.000000  0.000000  0.0  0.477121  0.778151  0.000000  0.000000   
d2  0.000000  0.778151  0.176091  0.0  0.000000  0.000000  0.000000  0.176091   
d3  0.000000  0.000000  0.088046  0.0  0.000000  0.000000  0.150515  0.088046   
d4  0.000000  0.000000  0.000000  0.0  0.477121  0.000000  0.301030  0.000000   
d5  0.778151  0.000000  0.176091  0.0  0.000000  0.000000  0.301030  0.176091   

     maullar   se         a    y     perro  despierta    duerme  después  \
d0  0.000000  0.0  0.000000  0.0  0.150515   0.000000  0.238561      0.0   
d1  0.000000  0.0  0.477121  0.0  0.301030   0.477121  0.000000      0.0   
d2  0.000000  0.0  0.000000  0.0  0.301030   0.000000  0.000000      0.0   
d3  0.000000  0.0  0.000000  0.0  0.000000   0.000000  0.238561      0.0   
d4  0.778151  0.0  0.477121  0.0  0.000000   0.477121  0.000000      0.0   
d5  0.000000  0.0  0.000000  0.0  0.000000   0.000000  0.000000      0.0   

        come  
d0  0.088046  
d1  0.000000  
d2  0.176091  
d3  0.088046  
d4  0.000000  
d5  0.176091

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [42]:
def cosine_similarity(v1, v2):
    return np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


> De la matriz de similitudes, se puede concuir que los documentos más similares son el documento *d0* ("El perro se come la comida y después se duerme") y el documento *d3* ("El gato se come la comida y después se duerme")

![gato](https://www.elagoradiario.com/wp-content/uploads/2020/05/Gato-mascarilla.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**